Java06. ДЗ 05, Егоров Антон

pvktk opened the pull request.

pvktk committed:
Начал писать сервер на асинхронных каналах

pvktk committed:
Написана часть кода для клиента.

pvktk committed:
Написан скачиватель отдельного файла.

pvktk committed:
Убран ReadWriteLock в клиенте. Мапы сделаны concurrent.
Ещё дописано команд в REPL.

pvktk committed:
Весь код написан.

pvktk committed:
Настроен maven-assembly-plugin

pvktk committed:
В main запускается один главный поток, для облегчения тестирования.

pvktk committed:
Добавлен тест для сервера.

pvktk committed:
Добавлена инструкция по сборке.

Сделаны join потоков сервера.

pvktk committed:
Исправлена бага при чтении из канала в буфер.

pvktk committed:
Исправлены баги сервера.

pvktk committed:
Добавлен тест реакции клиента на базовые команды.

pvktk committed:
Добавлен тест на скачивание с другого пира.

pvktk committed:
Добавлен тест скачивания с 2 пиров.

Добавлена команда resume.

pvktk committed:
Исправлена бага при повторном добавлении файла.

До этого для него выделялся новый id.

pvktk committed:
Переделал на RandomAccessFile.

pvktk committed:
Добавлено отображение размера файла в частях в выводе команды list.

pvktk committed:
Унифицирован код команд status и list.

pvktk committed:
обновлен readme

pvktk committed:
Подправлены сообщения о включении/выключении.

pvktk committed:
Добавлена обработка ввода чего-то вместо числа.

pvktk committed:
Сделано ограничение на количество одновременно скачивающихся кусков файла.

sproshev commented on torrent/src/main/java/torrent/client/FileProblemException.java:
```diff
@@ -0,0 +1,14 @@
+	/**
+	 * 
+	 */
+	private static final long serialVersionUID = 1L;
```
зачем? оно ж не сериализуется

pvktk replied:
Это подсказка IDE была.

sproshev commented on torrent/src/main/java/torrent/client/FilesDownloader.java:
```diff
@@ -0,0 +1,54 @@
+
+	ExecutorService pool = Executors.newCachedThreadPool();
+
+	Map<Integer, Future<?>> fileDownloadsFutures = new HashMap<>();
```
модификаторы доступа? не, не слышал)

sproshev replied:
если сделали старт, но не делали стоп, получается мапа будет содержать выполненные future сколь угодно долго?

pvktk replied:
Исправил.

sproshev commented on torrent/src/main/java/torrent/client/FilesDownloader.java:
```diff
@@ -0,0 +1,54 @@
+
+	public void stopFileDownload(int fileId) {
+		if (!filesHolder.fileStatus.containsKey(fileId)) {
+			throw new NullPointerException("This file wasn't been downloading");
```
это исключение про разыменование нулевого указателя, здесь скорее IllegalArgumentException/IllegalStateException/NoSuchElementException

sproshev commented on torrent/src/main/java/torrent/client/FilesHolder.java:
```diff
@@ -0,0 +1,193 @@
+		try {
+			return Math.toIntExact((files.get(fileId).length() + pieceSize - 1) / pieceSize);
+		} catch (IOException e) {
+			throw new RuntimeException(e);
```
можно еще java.io.UncheckedIOException

sproshev commented on torrent/src/main/java/torrent/client/FilesHolder.java:
```diff
@@ -0,0 +1,193 @@
+		}
+	}
+
+	public int pieceOffset(int fileId, int numPart) {
```
не используется

sproshev commented on torrent/src/main/java/torrent/client/FilesHolder.java:
```diff
@@ -0,0 +1,193 @@
+
+		for (Entry<Integer, String> ent : filePaths.entrySet()) {
+			try {
+				files.put(ent.getKey(), new RandomAccessFile(ent.getValue(), "rws"));
```
открытие дескриптора для всех них(

pvktk replied:
Сделал ленивым.

sproshev commented on torrent/src/main/java/torrent/client/FilesHolder.java:
```diff
@@ -0,0 +1,193 @@
+
+		try {
+			Paths.get(filePath).getParent().toFile().mkdirs();
+		} catch (NullPointerException npe) {}
```
откуда он ожидается, и почему исключение проигнорировано?

pvktk replied:
Это getParent() может null выдать.

sproshev replied:
разматывание стэка при создании исключения сильно дороже, чем `if (... != null)`

sproshev commented on torrent/src/main/java/torrent/client/FilesHolder.java:
```diff
@@ -0,0 +1,193 @@
+			throw new RuntimeException("Attemp to put block of incorrect size");
+		}
+		RandomAccessFile file = files.get(fileId);
+		file.seek(pieceId * pieceSize);
```
pieceOffset(...)

sproshev commented on torrent/src/main/java/torrent/client/MainInner.java:
```diff
@@ -0,0 +1,84 @@
+					out,
+					sinp);
+
+			replTh = new Thread(() -> repl.startRepl());
```
replTh = new Thread(repl::startRepl);

sproshev commented on torrent/src/main/java/torrent/client/MainInner.java:
```diff
@@ -0,0 +1,84 @@
+			}
+		} catch (IOException e) {
+			out.println(e.getMessage());
+		} catch (InterruptedException e) {
```
возможно лучше finally

sproshev commented on torrent/src/main/java/torrent/client/PieceDownloader.java:
```diff
@@ -0,0 +1,63 @@
+			} catch (IOException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+			}
```
не очень хорошая обработка IOException
семафор может быть не релизнут и/или канал не закрыт

sproshev commented on torrent/src/main/java/torrent/client/ServerFilesLister.java:
```diff
@@ -0,0 +1,46 @@
+			dout.writeByte(1);
+
+			fileSizes.clear();
+			filesNames.clear();
```
лучше все-таки не менять передаваемые параметры, а результат работы "возвращать"

pvktk replied:
Но тут параметров много.

sproshev replied:
и?) это же не оправдывает, в плюсах так исторически сложилось, что можно, в джаве скорее нет

sproshev commented on torrent/src/main/java/torrent/client/ServerFilesLister.java:
```diff
@@ -0,0 +1,46 @@
+				dinp.close();
+				dout.close();
+				s.close();
+			} catch (NullPointerException npe) {
```
это совсем беда, NPE это ошибка программирования, и оно никак не может оказаться в ожидаемых исключениях, вместо этого должна использоваться проверка на null

pvktk replied:
Тут могли не создаться, например, dinp и dout.

pvktk replied:
Кажется что это красивее чем делать три if и пытаться сообщить что кто-то оказался null.

sproshev replied:
сокет закрывает свои потоки

sproshev replied:
да и try-with-resources есть

pvktk replied:
Сделал вместо этого проверку на null.

sproshev commented on torrent/src/main/java/torrent/client/ServerFilesLister.java:
```diff
@@ -0,0 +1,46 @@
+			dout = new DataOutputStream(s.getOutputStream());
+			dinp = new DataInputStream(s.getInputStream());
+
+			dout.writeByte(1);
```
flush

sproshev reviewed (approved):
10

pvktk committed:
Исправил замечания.

В частности сделал загрузку файлового дискриптора ленивой
(только когда надо считать или записать кусок).

Future из мапы в FilesDownloader удаляется при завершении загрузки.

pvktk committed:
Сделал double check вместо synchronized в getOrCreateFile в FilesHolder.

Исправил catch NPE

pvktk committed:
Убраны закрытия DataStream привязанных к сокету.
